kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/KeyPair.java

```diff
@@ -39,4 +57,76 @@ public record KeyPair(@JsonProperty(required = true) String privateKeyFile,
     public <T> T accept(KeyProviderVisitor<T> visitor) {
         return visitor.visit(this);
     }
+
+    /**
+     * Test if this certificate has SANs matching a set of hostnames
+     * @param hostnames hostnames
+     * @return true if all hostnames match an SAN of this certificate
+     */
+    public boolean matchesHostnames(Set<String> hostnames) {
+        Set<Pattern> patterns = hostnamePatterns();
+        return hostnames.stream().allMatch(hostname -> patterns.stream().map(pattern -> pattern.matcher(hostname)).anyMatch(Matcher::matches));
+    }
+
+    private Set<Pattern> hostnamePatterns() {
+        try {
+            Path certPath = Path.of(certificateFile);
+            String pemString = Files.readString(certPath);
+            X509Certificate certificate = parsePemX509(pemString);
+            Set<String> dnsSubjectAlternativeNames = getDnsSubjectAlternativeNames(certificate);
+            // fallback to common name
+            if (dnsSubjectAlternativeNames.isEmpty()) {
+                String name = certificate.getSubjectX500Principal().getName();
+                if (name.startsWith("CN=")) {
+                    dnsSubjectAlternativeNames = Set.of(name.substring("CN=".length()));
+                }
+                else {
+                    throw new RuntimeException("failed to extract common name from certificate");
+                }
+            }
+            return dnsSubjectAlternativeNames.stream().map(KeyPair::toHostnamePattern).collect(toSet());
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    // need to handle wildcard (*.xyz) and literal (abc.xyz)
+    private static Pattern toHostnamePattern(String san) {
+        if (san.startsWith("*.")) {
```

Comment:
I believe hostnames are case insensitive, so I think the SAN/CN match should be case insensitive too.